packages/core/test/specs/undo_manager/index.ts

```diff
@@ -0,0 +1,296 @@
+import UndoManager from '../../../src/undo_manager';
+import Editor from '../../../src/editor';
+import { setupTestEditor } from '../../common';
+
+describe('Undo Manager', () => {
+  let editor: Editor;
+  let um: UndoManager;
+  let wrapper: any;
+
+  beforeEach(() => {
+    ({ editor, um } = setupTestEditor());
+    wrapper = editor.getWrapper();
+    um.clear();
+  });
+
+  afterEach(() => {
+    editor.destroy();
+  });
+
+  test('Initial state is correct', () => {
+    expect(um.hasUndo()).toBe(false);
+    expect(um.hasRedo()).toBe(false);
+    expect(um.getStack()).toHaveLength(0);
+  });
+
+  describe('Component changes', () => {
+    test('Add component', () => {
+      expect(wrapper.components()).toHaveLength(0);
+      wrapper.append('<div></div>');
+      expect(wrapper.components()).toHaveLength(1);
+      expect(um.hasUndo()).toBe(true);
+
+      um.undo();
+      expect(wrapper.components()).toHaveLength(0);
+      expect(um.hasRedo()).toBe(true);
+
+      um.redo();
+      expect(wrapper.components()).toHaveLength(1);
+    });
+
+    test('Remove component', () => {
+      const comp = wrapper.append('<div></div>')[0];
+      expect(wrapper.components()).toHaveLength(1);
+      um.clear();
+
+      comp.remove();
+      expect(wrapper.components()).toHaveLength(0);
+      expect(um.hasUndo()).toBe(true);
+
+      um.undo();
+      expect(wrapper.components()).toHaveLength(1);
+      expect(um.hasRedo()).toBe(true);
+
+      um.redo();
+      expect(wrapper.components()).toHaveLength(0);
+    });
+
+    test('Modify component properties', () => {
+      const comp = wrapper.append({ tagName: 'div', content: 'test' })[0];
+      um.clear();
+
+      comp.set('content', 'test2');
+      expect(comp.get('content')).toBe('test2');
+      expect(um.hasUndo()).toBe(true);
+
+      um.undo();
+      expect(comp.get('content')).toBe('test');
+
+      um.redo();
+      expect(comp.get('content')).toBe('test2');
+    });
+
+    test('Modify component style (StyleManager)', () => {
+      const comp = wrapper.append('<div></div>')[0];
+
+      setTimeout(() => {
+        um.clear();
+        comp.addStyle({ color: 'red' });
+        expect(comp.getStyle().color).toBe('red');
+        expect(um.hasUndo()).toBe(true);
+
+        um.undo();
+        expect(comp.getStyle().color).toBeUndefined();
+
+        um.redo();
+        expect(comp.getStyle().color).toBe('red');
+      });
+    });
```

Comment:
I think with setTimeout you should use `done`, otherwise it won't see failing asserts
```js
test('....', done => {
 // ...
 setTimeout(() => {
  // ...
  done(); 
 })
})
```